### PR TITLE
Update on conjugate gradient and its example

### PR DESCRIPTION
**Pull request by alineyyy**
This solves #249 .
CG is moved from extras module to a method. The corresponding test and example have also been updated.
This also aims to solve the first 2 issues in #218 .

### CODE REVIEW

File: src/mrinufft/operators/base.py [hunk 1]
```diff
@@ -365,6 +365,28 @@ def get_lipschitz_cst(self, max_iter=10, **kwargs):
...
+        computer_loss: bool
+            Whether to compute the loss at each iteration. 
+            If True, loss is calculated and returned, otherwise, it's skipped.
+        **kwargs:
+            Extra arguments to pass to the conjugate gradient method.
+
+        Returns
+        -------
+        np.ndarray, np.ndarray
+            Reconstructed image and loss value if compute_loss is True.
```
Comment (paquiteau) [inline src/mrinufft/operators/base.py]:
```suggestion
        np.ndarray
            Reconstructed image 
        np.ndarray, optional
            array of loss at each iteration, if compute_loss is True. 
```

**Review by paquiteau — CHANGES_REQUESTED** [review]
Small fixes but big implications, thank you ! 

Just a minor nitpick on the docstring

File: src/mrinufft/extras/gradient.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 
 
 @with_numpy
-def cg(operator, kspace_data, x_init=None, num_iter=10, tol=1e-4):
+def cg(operator, kspace_data, x_init=None, num_iter=10, tol=1e-4, compute_loss=False):
     """
     Perform conjugate gradient (CG) optimization for image reconstruction.
```
Comment (chaithyagr) [inline src/mrinufft/extras/gradient.py:12]: I think we can remove this if we have added its support in base.py already.

File: src/mrinufft/operators/base.py [hunk 3]
```diff
@@ -365,6 +365,32 @@ def get_lipschitz_cst(self, max_iter=10, **kwargs):
...
+        np.ndarray
+            Reconstructed image
+        np.ndarray, optional
+            array of loss at each iteration, if compute_loss is True.
+        """
+        from ..extras.gradient import cg
+
+        return cg(
+            operator=self, kspace_data=kspace_data, compute_loss=compute_loss, **kwargs
+        )
```
Comment (chaithyagr) [inline src/mrinufft/operators/base.py:392]: Hmm, this is new, i am okay with this also. @paquiteau thoughts?

File: src/mrinufft/extras/gradient.py [hunk 4]
```diff
@@ -47,16 +47,23 @@ def cg(operator, kspace_data, x_init=None, num_iter=10, tol=1e-4):
...
-            grad.flatten(), grad.flatten()
-        )
+        beta = np.dot(
+            grad_new.flatten(), (grad_new.flatten() - grad.flatten())
+        ) / np.dot(grad.flatten(), grad.flatten())
+        beta = max(0, beta)  # Polak-Ribiere formula is used to compute the beta
         velocity = grad_new + beta * velocity
 
         image = image - velocity / Lipschitz_cst
+        loss.append(calculate_loss(image)) if loss is not None else None
```
Comment (chaithyagr) [inline src/mrinufft/extras/gradient.py:67]:
```suggestion
        if compute_loss:
             loss.append(calculate_loss(image))
```

**Review by chaithyagr — CHANGES_REQUESTED** [review]
Thank you @alineyyy for this.

File: src/mrinufft/extras/gradient.py (hunk 2 above)
Comment (chaithyagr) [reply src/mrinufft/extras/gradient.py:12]: actually lets wait to see what @paquiteau thinks of having both a function and also the operator way to access this

File: src/mrinufft/operators/base.py (hunk 3 above)
Comment (paquiteau) [reply src/mrinufft/operators/base.py:392]: I am fine with this, having the `base.py` file remaining abstract and the implementation in the extra module is neat and tidy for me :)

File: examples/GPU/example_cg.py [hunk 5]
```diff
@@ -14,14 +14,9 @@
...
-References
-----------
-- Inpirations:
-        - https://sigpy.readthedocs.io/en/latest/_modules/sigpy/alg.html#ConjugateGradient
-        - https://aquaulb.github.io/book_solving_pde_mooc/solving_pde_mooc/notebooks/05_IterativeMethods/05_02_Conjugate_Gradient.html
-- Wikipedia:
-        - https://en.wikipedia.org/wiki/Conjugate_gradient_method
-        - https://en.wikipedia.org/wiki/Momentum
+This method is inspired by techniques from [SigPy]_ and
+[Aquaulb's MOOC]_, as well as general knowledge in [Wikipedia Conjugate Gradient]_.
```
Comment (chaithyagr) [inline examples/GPU/example_cg.py:18]:
```suggestion
[Aquaulb]_'s MOOC, as well as general knowledge in [Wikipedia Conjugate Gradient]_.
```

File: examples/GPU/example_cg.py [hunk 6]
```diff
@@ -60,21 +55,75 @@
 # %%
 # Reconstruct the image using the CG method
 kspace_data = nufft.op(image)  # get the k-space data
-reconstructed_image = cg(nufft, kspace_data)  # reconstruct the image
+reconstructed_image, loss = nufft.cg(
+    kspace_data=kspace_data, num_iter=50, compute_loss=True
+)  # reconstruct the image
```
Comment (chaithyagr) [inline examples/GPU/example_cg.py:60]:
```suggestion
)
```

File: examples/GPU/example_cg.py [hunk 7]
```diff
@@ -60,21 +55,75 @@
...
-plt.imshow(abs(reconstructed_image), cmap="gray")
+plt.imshow(
+    normalize(
+        abs(reconstructed_image),
+        abs(reconstructed_image).min(),
+        abs(reconstructed_image).max(),
+    ),
+    cmap="gray",
+)
+plt.colorbar()
```
Comment (chaithyagr) [inline examples/GPU/example_cg.py:84]: You dont need to `normalize` your data. Just ensure that vmax and vmin is same in all the cases.
```
    plt.imshow(image, cmap='gray')
    plt.imshow(reconstructed_image, vmax=image.max(), vmin=image.min(), cmap='gray')
```

Do the same for DC Adjoint.

File: examples/GPU/example_cg.py [hunk 8]
```diff
@@ -60,21 +55,75 @@
 # %%
 # Reconstruct the image using the CG method
 kspace_data = nufft.op(image)  # get the k-space data
-reconstructed_image = cg(nufft, kspace_data)  # reconstruct the image
+reconstructed_image, loss = nufft.cg(
+    kspace_data=kspace_data, num_iter=50, compute_loss=True
```
Comment (chaithyagr) [inline examples/GPU/example_cg.py:59]:
```suggestion
    kspace_data=kspace_data, x_init=nufft.adj_op(kspace_data), num_iter=50, compute_loss=True
```

File: examples/GPU/example_cg.py [hunk 9]
```diff
@@ -60,21 +55,75 @@
...
+
+plt.plot(kspace_data, label="acquired kspace")
+plt.plot(nufft.op(reconstructed_image), alpha=0.7, label="reconstructed kspace")
+plt.legend(loc="lower left", fontsize=8)
+
+plt.subplot(2, 3, 6)
+plt.title("kspace from adjoint NUFFT")
+plt.plot(kspace_data, label="acquired kspace")
+plt.plot(nufft.op(image), alpha=0.7, label="reconstructed kspace")
+plt.legend(loc="lower left", fontsize=8)
```
Comment (chaithyagr) [inline examples/GPU/example_cg.py:114]: Perhaps we should plot this on log scale. Plot it as np.log(np.abs(kspace_data))

This will help better showcase the difference.

File: src/mrinufft/extras/gradient.py [hunk 10]
```diff
@@ -47,16 +47,23 @@ def cg(operator, kspace_data, x_init=None, num_iter=10, tol=1e-4):
     velocity = tol * velocity + grad / Lipschitz_cst
```
Comment (chaithyagr) [inline src/mrinufft/extras/gradient.py:47]:
```suggestion
    velocity = tol * velocity + grad / lipschitz_cst
```

Please make the same changes bellow..

Comment (chaithyagr) [reply src/mrinufft/extras/gradient.py:47]: The variable should start with short letter.

**Review by chaithyagr — CHANGES_REQUESTED** [review]
Second round of revision. Getting nitpicky here and there..
